### PR TITLE
Fix datatype of some TM. Partial fix #1576

### LINKED ISSUE #1576: Check what's a reason to use double-precision in TMs
Our TMs return vectors with double-precision `float64`, it looks like very suspicious, because `float32` is enough for all. Need to check, what's a reason of this behavior and what's a concrete method.

The first step - look at [this line in the test](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/basetests.py#L51), after it - collect all TMs, that depends on this tests and check, where and why `float64` happened.

Result - detailed description (where and why), and fixing this behavior after discussion (if needed)

### PR DESCRIPTION
**Pull request by Yuktichopra**

### CODE REVIEW

File: gensim/models/atmodel.py [hunk 1]
```diff
@@ -65,11 +65,12 @@ class AuthorTopicState(LdaState):
...
-    def __init__(self, eta, lambda_shape, gamma_shape):
-        self.eta = eta
-        self.sstats = np.zeros(lambda_shape)
-        self.gamma = np.zeros(gamma_shape)
+    def __init__(self, eta, lambda_shape, gamma_shape,dtype=np.float64):
+        self.eta = eta.astype(dtype, copy=False)
+        self.sstats = np.zeros(lambda_shape, dtype)
+        self.gamma = np.zeros(gamma_shape, dtype)
         self.numdocs = 0
+        self.dtype = np.float64
```
Comment (menshikh-iv) [inline gensim/models/atmodel.py]: probably you should use `dtype` that passed as parameter

File: gensim/models/atmodel.py [hunk 2]
```diff
@@ -65,11 +65,12 @@ class AuthorTopicState(LdaState):
 
     """
 
-    def __init__(self, eta, lambda_shape, gamma_shape):
-        self.eta = eta
-        self.sstats = np.zeros(lambda_shape)
-        self.gamma = np.zeros(gamma_shape)
+    def __init__(self, eta, lambda_shape, gamma_shape,dtype=np.float64):
```
Comment (menshikh-iv) [inline gensim/models/atmodel.py]: `,dt` -> `, dt`

File: gensim/models/hdpmodel.py [hunk 3]
```diff
@@ -126,7 +127,7 @@ class HdpModel(interfaces.TransformationABC, basemodel.BaseTopicModel):
     def __init__(self, corpus, id2word, max_chunks=None, max_time=None,
                  chunksize=256, kappa=1.0, tau=64.0, K=15, T=150, alpha=1,
                  gamma=1, eta=0.01, scale=1.0, var_converge=0.0001,
-                 outputdir=None, random_state=None):
+                 outputdir=None, random_state=None, dtype = float32):
```
Comment (menshikh-iv) [inline gensim/models/hdpmodel.py]: maybe np.float32?

File: gensim/models/ldaseqmodel.py [hunk 4]
```diff
@@ -421,7 +422,7 @@ def __getitem__(self, doc):
         """
         lda_model = ldamodel.LdaModel(num_topics=self.num_topics, alpha=self.alphas, id2word=self.id2word)
         lda_model.topics = np.array(np.split(np.zeros(self.vocab_len * self.num_topics), self.vocab_len))
-        ldapost = LdaPost(num_topics=self.num_topics, max_doc_len=len(doc), lda=lda_model, doc=doc)
+        ldapost = LdaPost(num_topics=self.num_topics, max_doc_len=len(doc), lda=lda_model, doc=doc, dtype = float32)
```
Comment (menshikh-iv) [inline gensim/models/ldaseqmodel.py]: why float32?

File: gensim/models/atmodel.py [hunk 5]
```diff
@@ -67,12 +67,16 @@ class AuthorTopicState(LdaState):
...
+        self.eta = eta.astype(dtype, copy=False)
+        self.sstats = np.zeros(lambda_shape, dtype)
+        self.gamma = np.zeros(gamma_shape, dtype)
         self.numdocs = 0
-        self.dtype = np.float64  # To be compatible with LdaState
+
+        self.dtype = np.float64
+
+        self.dtype = np.float32  # To be compatible with LdaState
```
Comment (menshikh-iv) [inline gensim/models/atmodel.py]: What happens here (line 76 and 78)?

File: gensim/models/atmodel.py [hunk 6]
```diff
@@ -207,9 +211,13 @@ def __init__(self, corpus=None, num_topics=100, id2word=None, author2doc=None, d
         ... corpus, num_topics=50, author2doc=author2doc, id2word=id2word, alpha='auto', eval_every=5)
 
         """
+
+        self.dtype = np.float64
+=======
```
Comment (menshikh-iv) [inline gensim/models/atmodel.py]: oh, looks like bad resolving of merge conflict (here and everywhere)

**Comment by Yuktichopra** [comment]
I did not understand this, please elaborate. For this model to be
compatible with ldamodel, and the class that is called in each of these
models, shoudn't I change it to float 64?

File: gensim/models/atmodel.py (hunk 5 above)
Comment (Yuktichopra) [reply gensim/models/atmodel.py]: My bad

**Comment by menshikh-iv** [comment]
@Yuktichopra please make dtype configurable now (probably later we'll "pin" it), don't pin to `float64` or another type.

Also, check properly than dtype isn't change (tests + probably asserts in code)

**Comment by Yuktichopra** [comment]
Shall I look for backward compatibility?

**Comment by menshikh-iv** [comment]
@Yuktichopra of course!

**Comment by menshikh-iv** [comment]
Ping @Yuktichopra, how is going?

**Comment by menshikh-iv** [comment]
PR looks abandoned, for this reason, I close it.

**Pull request closed without merging**